File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/RksResult.java
package generations.gg.generations.core.generationscore.common.world.recipe;

import com.google.gson.JsonElement;
import com.mojang.serialization.Codec;
import com.mojang.serialization.JsonOps;
import generations.gg.generations.core.generationscore.common.world.level.block.entities.RksMachineBlockEntity;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;

import java.util.function.Function;

public interface RksResult<U extends RksResult<U>> {
    Codec<RksResult<?>> CODEC = RksResultType.CODEC.dispatch(RksResult::type, RksResultType::codec);

    ItemStack getStack();

    RksResultType<U> type();

    default void process(Player player, RksMachineBlockEntity rksMachineBlockEntity, ItemStack stack) {}

    default boolean consumeTimeCapsules() {
        return true;
    }

    boolean isPokemon();

    default void toBuffer(FriendlyByteBuf buffer) {
        buffer.writeResourceLocation(RksResultType.RKS_RESULT.getId(type()));
        type().toBuffer().accept(buffer, (U) this);
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/common/world/recipe/RksResult.kt
package generations.gg.generations.core.generationscore.common.world.recipe

import com.mojang.serialization.Codec
import generations.gg.generations.core.generationscore.common.world.level.block.entities.RksMachineBlockEntity
import generations.gg.generations.core.generationscore.common.world.recipe.RksResultType
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import java.util.function.Function

interface RksResult<U : RksResult<U>> {
    val stack: ItemStack

    fun type(): RksResultType<U>

    fun process(player: Player, rksMachineBlockEntity: RksMachineBlockEntity, stack: ItemStack) {}

    fun consumeTimeCapsules(): Boolean {
        return true
    }

    val isPokemon: Boolean

}
